include/linux/vfs_compat.h

```diff
@@ -352,6 +353,48 @@ static inline struct inode *file_inode(const struct file *f)
 }
 #endif /* HAVE_FILE_INODE */
 
+#ifdef HAVE_KUID_HELPERS
+static inline uid_t zfs_uid_read(struct inode *ip)
+{
+	return (from_kuid(kcred->user_ns, ip->i_uid));
+}
+
+static inline uid_t zfs_gid_read(struct inode *ip)
+{
+	return (from_kgid(kcred->user_ns, ip->i_gid));
+}
+
+static inline void zfs_uid_write(struct inode *ip, uid_t uid)
+{
+	ip->i_uid = SUID_TO_KUID(uid);
+}
+
+static inline void zfs_gid_write(struct inode *ip, gid_t gid)
+{
+	ip->i_gid = SGID_TO_KGID(gid);
```

Comment:
Why do the _read() go through ns translation, but not the _write()? We should stick to either one or the other.